File: CombatCore/FixedVector3.cs
namespace CombatCore;

public readonly struct FixedVector3 : IEquatable<FixedVector3>
{
    public readonly Fixed X, Y, Z;

    public FixedVector3(Fixed x, Fixed y, Fixed z)
    {
        X = x; Y = y; Z = z;
    }

    public static readonly FixedVector3 Zero = new(Fixed.Zero, Fixed.Zero, Fixed.Zero);

    public static FixedVector3 FromFloats(float x, float y, float z) =>
        new(Fixed.FromFloat(x), Fixed.FromFloat(y), Fixed.FromFloat(z));

    public static FixedVector3 operator +(FixedVector3 a, FixedVector3 b) =>
        new(a.X + b.X, a.Y + b.Y, a.Z + b.Z);

    public static FixedVector3 operator -(FixedVector3 a, FixedVector3 b) =>
        new(a.X - b.X, a.Y - b.Y, a.Z - b.Z);

    public static FixedVector3 operator -(FixedVector3 a) => new(-a.X, -a.Y, -a.Z);

    public static FixedVector3 operator *(FixedVector3 a, Fixed s) =>
        new(a.X * s, a.Y * s, a.Z * s);

    public static FixedVector3 operator *(FixedVector3 a, int s) =>
        new(a.X * s, a.Y * s, a.Z * s);

    public Fixed LengthSquared() => X * X + Y * Y + Z * Z;

    public Fixed Length() => Fixed.Sqrt(LengthSquared());

    public bool Equals(FixedVector3 other) => X == other.X && Y == other.Y && Z == other.Z;
    public override bool Equals(object? obj) => obj is FixedVector3 v && Equals(v);
    public override int GetHashCode() => HashCode.Combine(X, Y, Z);
    public override string ToString() => $"({X}, {Y}, {Z})";
}

File: Client/src/sim/FixedVector3.ts
import { Fixed } from "./Fixed.ts";

export class FixedVector3 {
  readonly x: Fixed;
  readonly y: Fixed;
  readonly z: Fixed;

  constructor(x: Fixed, y: Fixed, z: Fixed) {
    this.x = x;
    this.y = y;
    this.z = z;
  }

  static readonly ZERO = new FixedVector3(Fixed.ZERO, Fixed.ZERO, Fixed.ZERO);

  static fromFloats(x: number, y: number, z: number): FixedVector3 {
    return new FixedVector3(Fixed.fromFloat(x), Fixed.fromFloat(y), Fixed.fromFloat(z));
  }

  static add(a: FixedVector3, b: FixedVector3): FixedVector3 {
    return new FixedVector3(Fixed.add(a.x, b.x), Fixed.add(a.y, b.y), Fixed.add(a.z, b.z));
  }

  static sub(a: FixedVector3, b: FixedVector3): FixedVector3 {
    return new FixedVector3(Fixed.sub(a.x, b.x), Fixed.sub(a.y, b.y), Fixed.sub(a.z, b.z));
  }

  static scale(a: FixedVector3, s: Fixed): FixedVector3 {
    return new FixedVector3(Fixed.mul(a.x, s), Fixed.mul(a.y, s), Fixed.mul(a.z, s));
  }

  static scaleInt(a: FixedVector3, s: number): FixedVector3 {
    return new FixedVector3(Fixed.mulInt(a.x, s), Fixed.mulInt(a.y, s), Fixed.mulInt(a.z, s));
  }

  equals(other: FixedVector3): boolean {
    return this.x.equals(other.x) && this.y.equals(other.y) && this.z.equals(other.z);
  }

  toString(): string {
    return `(${this.x}, ${this.y}, ${this.z})`;
  }
}
